lib/plugin-manager.js

```diff
@@ -0,0 +1,302 @@
+    this._boundInit = this._init.bind(this);
+  }
+
+  /**
+   * Initialize the PluginManager's index and metadata source
+   */
+  initialize() {
+    Log.info('Initializing index and metadata');
+    this._running = true;
+    this._init();
+  }
+
+  /**
+   * Shutdown the plugin manager's sources and then the plugin manager
+   * @param {Boolean} del
+   */
+  shutdown(del) {
+    Log.info('Shutting down sources');
+    this.storage.sources.forEach((source) => {
+      source.shutdown();
+    });
+
+    if (del === true) {
+      this.storage.sources = [];
+    }
+
+    this._running = false;
+  }
+
+  /**
+   * Get the plugin manager's status
+   * @returns {{running: boolean, ok: boolean, sources: Array}}
+   */
+  status() {
+    return {
+      running: this._running,
+      ok: this._ok,
+      sources: this._sourcesStatus()
+    };
+  }
+
+  /**
+   * Gets an array of source statuses
+   * @returns {Array}
+   * @private
+   */
+  _sourcesStatus() {
+    return this.storage.sources.map((source) => {
+      return {
+        name: source.name,
+        type: source.type,
+        status: source.status()
+      };
+    });
+  }
+
+  /**
+   * Coordinates synchronizing the index and metadata sources and then mapping them together
+   * @private
+   */
+  _init() {
+    const boundIndex = this._index.bind(this);
+    const boundMapper = this._mapper.bind(this);
+
+    async.series([boundIndex, boundMapper], (err, response) => {
+      // There's no possible error condition because both this._mapper() and this._index() handle
+      // their own errors in order to retry their respective operations.
+      this._ok = true;
+
+      const sources = response[1];
+
+      this.emit('sources-generated', sources);
+      this._registerSources(sources);
+      this.emit('sources-registered', this.storage.sources);
+    });
+  }
+
+  /**
+   * Coordinate fetching the PluginManager's index and metadata source
+   * @param {Function} callback
+   * @returns {null}
+   * @private
+   */
+  _index(callback) {
+    return async.parallel([
+      this._createParallelTask(this.index),
+      this._createParallelTask(this.metadata)
+    ], (err, errSources) => {
+      if (err) {
+        // If any of the parallel tasks fail it means either the Metadata plugin or the index plugin has failed
+        // so we need to determine which plugin threw the error then set a listener on it for its next update to
+        // try to assemble the whole thing again.
+        this._error(err);
+
+        errSources.filter(Boolean).forEach((el) => {
+          el.shutdown();
+          el.once('update', this._boundInit);
+          el.initialize();
+        });
+        return;
+      }
+      callback(null);
+    });
+  }
+
+  /**
+   * Maps instance metadata into the index document
+   * @param {Function} callback
+   * @private
+   */
+  _mapper(callback) {
+    let sources = [];
+
+    try {
+      sources = this.index.properties.sources.map((el) => {
```

Comment:
I should clarify that the index source isn't being registered with the storage layer by design. It's a special case with a different syntax from the rest of the S3 source documents.